middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,196 @@
+
+        return getattr(
+            sys.modules[__name__],
+            PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[payload_type]
+        )(payload, request_id)
+
+    @classmethod
+    def get_message_format(cls) -> Struct:
+        message_format = Struct(
+            frame_signature=Bytes(64),  # TODO: How we gonna serialize
+            signed_part_of_the_frame=Struct(
+                request_id=Int32ub,
+                payload_length=Int16ub,
+                payload_type=Enum(Byte, PayloadType),
+                payload=Prefixed(VarInt, GreedyBytes),
+            ),
+            frame_separator=Const(FRAME_SEPARATOR),
+        )
+        return message_format
+
+    @classmethod
+    def deserialize(cls, raw_message):
+        message_format = cls.get_message_format()
+
+        deserialized_message = message_format.parse(raw_message)
+
+        cls._validate_signature(deserialized_message)
+        cls._validate_length(deserialized_message)
+
+        deserialized_payload = getattr(
+            sys.modules[__name__],
+            PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[
+                PayloadType[str(deserialized_message.signed_part_of_the_frame.payload_type)]
+            ]
+        )._deserialize_payload(deserialized_message.signed_part_of_the_frame.payload)
+
+        return deserialized_payload
+
+    @classmethod
+    def _deserialize_payload(cls, payload):
+        return payload
+
+    @classmethod
+    def _validate_length(cls, message):
+        if not len(message.signed_part_of_the_frame.payload) == message.signed_part_of_the_frame.payload_length:
+            raise MiddlemanProtocolError(
+                f'Deserialized message payload length is {len(message.signed_part_of_the_frame.payload)} '
+                f'instead of {message.signed_part_of_the_frame.payload_length}.'
+            )
+
+    @classmethod
+    def _validate_signature(cls, message):
+        if not message.frame_signature:  # TODO: Really verify signature
+            raise MiddlemanProtocolError(
+                f'Deserialized message signature wrong.'
+            )
+
+    def serialize(self) -> bytes:
+        assert isinstance(self.payload, bytes)
+
+        message_format = self.get_message_format()
+
+        signed_part_of_the_frame = Container(
+            request_id=self.request_id,
+            payload_length=len(self.payload),
+            payload_type=self.payload_type,
+            payload=self.payload,
+        )
+
+        raw_signed_part_of_the_frame = message_format.signed_part_of_the_frame.build(
+            signed_part_of_the_frame
+        )
+
+        #frame_signature = sha256(raw_signed_part_of_the_frame).hexdigest().encode()  # TODO: TMP
+
+        frame_signature = b'x' * 64
+
+        message = Container(
+            frame_signature=frame_signature,
+            signed_part_of_the_frame=signed_part_of_the_frame,
+        )
+
+        raw_message = message_format.build(message)
+        return raw_message
```

Comment:
You're not escaping the raw message to ensure that it never contains the separator.

Reply:
:+1: done